Add modeLogic and mode link UI support

dthubereng opened the pull request:
Fixes #1270 

Add MSP_MODE_RANGES_EXTRA to msp codes
Add MODE_RANGES_EXTRA function call
Add modeRangeExtra when reading mode settings
Add UI for modeLogic
Position and populate Logic selector
Some fixes for msp apiVersion checking
Add 'save' modeLogic
Add link button and translations
Add code to show/hide Link button
Adjust position of link and range buttons
Add structure code for Link button
Add linkedTo code
Add link div with delete button to html and css.
Add code to hide/show buttons when link is pressed.
Add code to add link select to range area.
Add save code for linked modes
Add formatting to link
Don't allow linking to/from ARM or self
ModeId changes

mikeller commented on src/css/tabs/auxiliary.css:
```diff
@@ -127,15 +145,20 @@
-.tab-auxiliary .mode .range .channelInfo .limits {
-    padding: 10px 0px;
-}
+/*.tab-auxiliary .mode .range .channelInfo .limits {
```
If this is not used any more please remove it.

dthubereng replied:
Being unfamiliar with css, I wasn't sure. It was an artifact of testing things out and didn't seem to have any affect on any thing. I will remove.

mikeller commented on locales/ja/messages.json:
```diff
@@ -1543,6 +1543,9 @@
         "message": "範囲追加"
     },
+    "auxiliaryAddLink": {
+        "message": "結ぶ"
```
Wow, I am impressed.
For next time, adding the translations (besides English) is not required in a pull request. We have a team of volunteer translators (hanging out in the #team_translation channel in Slack), and they use crowdin (https://crowdin.com/) to translate the messages, and the non-English message files are generated from there.

McGiverGim replied:
Yes, they will be overwritten from there before the release of the next version. Sorry for the extra work ;)

dthubereng replied:
No worries. I've always loved linguistics and this was an enjoyable exercise, even if it does get overwritten. It was a welcome break from trying to translate/speak javascript. :)

mikeller commented on src/js/msp/MSPHelper.js:
```diff
@@ -785,6 +785,37 @@ MspHelper.prototype.process_data = function(dataHandler) {
                 }
+
+                // MODE_RANGES_EXTRA isn't support before 1.42
+                if (semver.lt(CONFIG.apiVersion, "1.42.0")){
```
The plan is to add this for 4.0, which will be API version 1.41.

Also, please use a space before `{`.

I probably wouldn't stick filling in the defaults into the processing of `MSP_MODE_RANGES`, this is a bit unintuitive. If the adjustment range tab needs default values for `MODE_RANGES_EXTRA`, then the tab should really supply them if `MSP_MODE_RANGES_EXTRA` isn't called.

dthubereng replied:
Ah, I get it. Should have checked that the current release was 1.40.

dthubereng replied:
And I think you meant auxiliary tab. Yes, it needs default values of `MSP_MODE_RANGES`, and after having worked in the tab code a bit now, I think I have a better understanding of what I should have done. I'll make those changes.

dthubereng replied:
If I'm reading the code right, it looks like `MODE_RANGES_EXTRA` support needs to be added to backup_restore.js, correct? I can see a call there that would also require default values if the api doesn't support it. I'll test this.

mikeller replied:
Yes please, add it to `backup_restore.js`.

dthubereng replied:
@mikeller: I added default code to `backup_restore.js`. However, I did not make changes to `migrate()` in `backup_restore.js`. It looked like there were some issues with it and didn't look like it had been updated in a while.

mikeller commented on src/js/msp/MSPHelper.js:
```diff
@@ -785,6 +785,37 @@ MspHelper.prototype.process_data = function(dataHandler) {
+            case MSPCodes.MSP_MODE_RANGES_EXTRA:
+                MODE_RANGES_EXTRA = []; // empty the array as new data is coming in
+
+                var modeRangeCount = (data.byteLength - 1) / data.readU8();
```
This will need changing after https://github.com/betaflight/betaflight/pull/7543#discussion_r254636886 is done.

dthubereng replied:
👍

mikeller commented on src/js/msp/MSPHelper.js:
```diff
@@ -1913,19 +1944,21 @@ MspHelper.prototype.sendModeRanges = function(onCompleteCallback) {
     function send_next_mode_range() {
 
         var modeRange = MODE_RANGES[modeRangeIndex];
+        var modeRangeExtra = MODE_RANGES_EXTRA[modeRangeIndex];
```
This is a bit confusing here - you are sending `permanentId` as as part of `MSP_MODE_RANGE_EXTRA`, but then you are not using it here to match entries for the two arrays.

I think the match by index that you are doing is a correct way of doing it, but there should be a remark in `msp.c` in betaflight/betaflight#7543 that stipulates that the elements of `MSP_MODE_RANGE` and `MSP_MODE_RANGE_EXTRA` be aligned.

dthubereng replied:
I understand the confusion. I'll add the comment in `msp.c` in [betaflight/betaflight#7543](https://github.com/betaflight/betaflight/pull/7543).

mikeller commented on src/js/tabs/auxiliary.js:
```diff
@@ -8,7 +8,12 @@ TABS.auxiliary.initialize = function (callback) {
     function get_mode_ranges() {
-        MSP.send_message(MSPCodes.MSP_MODE_RANGES, false, false, get_box_ids);
+        MSP.send_message(MSPCodes.MSP_MODE_RANGES, false, false, 
+            semver.gte(CONFIG.apiVersion, "1.42.0") ? get_mode_ranges_extra : get_box_ids);
```
Again `1.41.0`.

dthubereng replied:
👍

mikeller commented on src/js/tabs/auxiliary.js:
```diff
@@ -70,9 +81,50 @@ TABS.auxiliary.initialize = function (callback) {
+        logicOption.val(0);
+        logicList.append(logicOption);
+        
+        if(semver.gte(CONFIG.apiVersion, "1.42.0")){
```
Again 1.41.0.

dthubereng replied:
👍

mikeller commented:
There are a couple of problems in the UI left as well:

![image](https://user-images.githubusercontent.com/4742747/52410591-a3a4d980-2b3d-11e9-93c1-f9c750edfb03.png)

- the linked to mode is not shown after loading;
- I think it is confusing as to what the 'AND' and 'OR' dropdowns do, and how they are applied. I think there should be a tooltip explaining this, and they should be hidden when they are not used (i.e. one range for a mode only).

dthubereng commented:
I've addressed the code comments and will look at the UI issue/suggestion next.

@mikeller: I also thought about the tooltip and hide/show logic and think it's a great idea. As for the link mode not being shown, I think I know how you may have gotten that and will look at addressing that case.

mikeller commented on src/js/backup_restore.js:
```diff
@@ -34,6 +34,9 @@ function configuration_backup(callback) {
         if (semver.gte(CONFIG.apiVersion, "1.15.0")) {
             profileSpecificData.push(MSPCodes.MSP_RC_DEADBAND);
         }
+        if (semver.gte(CONFIG.apiVersion, "1.41.0")) {
```
Looks like this has landed in the wrong block - mode ranges are not profile specific.

dthubereng replied:
ok, will have a look at that.

dthubereng replied:
@mikeller: would this then be true of `MSP_MODE_RANGES` defined in the original `profileSpecificData` on line 26? That's why I added EXTRAS here.

mikeller replied:
I see. The background to this is that in the early days, what is now known as 'PID profiles' was called 'profiles', and included a lot more settings, including mode ranges (probably originally intended to allow a user to use the same flight controller on different models). At this stage, 'profiles' included modes. `MSP_MODE_RANGES` was left in the profile specific section ever since - this causes the same mode range to be written thrice, but it retains backwards compatibility. For `MSP_MODE_RANGES_EXTRA`, there is no need to do the same, as there are no backwards compatibility considerations.

dthubereng replied:
ok, good to know.

dthubereng replied:
So, is there going to be a problem with the restore function? The same call to `MSP_SET_MODE_RANGES` contains the EXTRA data, too. I think `MODE_RANGES_EXTRA` will have to be read three times to conserve backwards compatibility.

mikeller replied:
Heh, good point, there is no practical downside to having `MODE_RANGES_EXTRA` written / read three times, so let's go for that.

dthubereng replied:
Found a way around loading `MODE_RANGES_EXTRA` three times.

mikeller commented on src/js/msp/MSPHelper.js:
```diff
@@ -1913,19 +1928,21 @@ MspHelper.prototype.sendModeRanges = function(onCompleteCallback) {
             .push8((modeRange.range.start - 900) / 25)
-            .push8((modeRange.range.end - 900) / 25);
+            .push8((modeRange.range.end - 900) / 25)
+            .push8(modeRangeExtra.modeLogic)
```
This should have a version check, and only append the extra parameters if the API version is >= 1.41.

dthubereng replied:
You're right. I thought betaflight handled it already, but it does make sense to also handle it here; it would be more robust.

dthubereng commented:
Doh! I forgot to add the tooltip. I'll add the tooltip to the AND/OR dropdown as well as add some description to the top of the page.

Also, I missed the mixing of links and ranges, so will add that as well. Stay tuned!

dthubereng commented:
Changed the code (simplified it, really) to be able to mix and match links and ranges. Still working on the Help and Tooltip documentation, partially because I'm trying to wrap my head around how it works now. Along those lines, there appears to be an issue with using the logical operators with the linked modes. So, I'll continue investigating that.

dthubereng commented:
@mikeller: to address your comment at https://github.com/betaflight/betaflight-configurator/pull/1287#pullrequestreview-201878752, it appears the firmware does not currently support a mixture of links and ranges currently. If a link is present, then it takes precedence and all of the ranges in the same mode are subsequently ignored. In fact, if there are multiple links, only the last link assigned to a mode has full control of the mode. At this point, I think the UI should support the current implementation and another PR generated with an improvement of the mode logic in the firmware to allow mixing links and ranges. In my opinion then, tooltips and Help text documenting mode logic would be updated with the revamped logic. Advice or comments?

mikeller commented:
@dthubereng: In this case I misled you, my apologies. I only tested that a range and a link can be assigned concurrently in CLI (they can), but not that they actually work (i.e. the CLI is broken by allowing them to be stored).
There is a number of things we can do:
1. Fix the CLI to throw if an illegal combination of range and link, or multiple links for the same mode are set. Leave configurator as it was in your last iteration;
2. Fix modes to actually support combinations, adapt configurator to this.
As the feature freeze for 4.0 is only a couple of weeks away, I think going for 1. for 4.0 is the safe bet. We can always look into doing 2. as a follow up after 4.0 if we think it will give value to users.

dthubereng commented:
@mikeller: I think 2 is not a difficult change; that is, changing the firmware to allow mixing links and ranges to activate a mode would not be difficult or invasive. I think mode logic needs to be revamped, but that's a pretty major change that should wait until after 4.0. I recommend letting me change the firmware and leaving Configurator changes with my latest commit.

dthubereng commented:
@mikeller: I believe this one is ready now. Let me know if you think the help text is sufficient. It will need to be updated if/when the logic is updated. If logic is not updated to be more intuitive, then the help text will have to be fleshed out more. Everything else should be good to go (backwards compatible, etc.).

mikeller commented:
This looks good now.

Unfortunately, when testing it, I found that the logic that is applied when combining modes is now wrong (in the firmware, with latest `master`): In the presence of one or more `AND` ranges for a mode, the `OR` ranges now evaluate to `AND NOT`:

![image](https://user-images.githubusercontent.com/4742747/53334467-cc123d80-395d-11e9-95ee-71d8060a369e.png)

![image](https://user-images.githubusercontent.com/4742747/53334483-db918680-395d-11e9-9fe4-5a34a35ee710.png)

Also, I think the part of the help text dealing with logic combinations should probably be rephrased so that users not versed in logic can understand it better:

> Multiple ranges/links in a single mode are by default logically OR'd together, but may also be logically AND'd together.

How about:

`If there is more than one range / link defined for a switch, each of them can be set to `AND` or `OR`. A switch will be on whenever:
- all `AND` ranges / links are on; and
- at least one of the `OR` ranges / links is on.'

dthubereng commented:
i actually noticed the same thing when testing initially. I think perhaps this is how the logic is currently implemented (I did not change that) and until we had a way of seeing it visually, it wasn’t intuitive. The way you have suggested writing the help text is how I think it should operate as well and was planning on addressing that after 4.0 is released. The dilemma now is do I make the help text the way it should work or change it after the firmware is updated (or a bit of both)?

dthubereng commented:
@mikeller: I went ahead and revamped the mode logic on the firmware. It's ready to be tested, but I have to wait until I get back to my computer in Texas (Wed or Thurs) to test with this latest configurator change (it's just easier and I can't gulp from this PC). If it works as expected, I'll PR that and then update the help text on this one; unless, of course, I hear from you to hold off.

mikeller commented:
@dthubereng: No, that's all good, looking forward to seeing your logic changes.
